x/quarantine/keeper/keeper.go

```diff
@@ -0,0 +1,424 @@
+		Coins: sdk.Coins{},
+	}
+	if len(bz) > 0 {
+		err := k.cdc.Unmarshal(bz, &qf)
+		if err != nil {
+			return &qf, err
+		}
+	}
+	return &qf, nil
+}
+
+// mustBzToQuarantineRecord returns bzToQuarantineRecord but panics on error.
+func (k Keeper) mustBzToQuarantineRecord(bz []byte) *quarantine.QuarantineRecord {
+	qf, err := k.bzToQuarantineRecord(bz)
+	if err != nil {
+		panic(err)
+	}
+	return qf
+}
+
+// GetQuarantineRecord gets the single quarantine record to toAddr from all the fromAddrs.
+// If the record doesn't exist, nil is returned.
+//
+// If you want all records from any of the fromAddrs, use GetQuarantineRecords.
+func (k Keeper) GetQuarantineRecord(ctx sdk.Context, toAddr sdk.AccAddress, fromAddrs ...sdk.AccAddress) *quarantine.QuarantineRecord {
+	store := ctx.KVStore(k.storeKey)
+	key := quarantine.CreateRecordKey(toAddr, fromAddrs...)
+	if store.Has(key) {
+		bz := store.Get(key)
+		qr := k.mustBzToQuarantineRecord(bz)
+		return qr
+	}
+	return nil
+}
+
+// GetQuarantineRecords gets all the quarantine records to toAddr that involved any of the fromAddrs.
+//
+// If you want a single record from all the fromAddrs, use GetQuarantineRecord.
+func (k Keeper) GetQuarantineRecords(ctx sdk.Context, toAddr sdk.AccAddress, fromAddrs ...sdk.AccAddress) []*quarantine.QuarantineRecord {
+	store := ctx.KVStore(k.storeKey)
+	allSuffixes := k.getQuarantineRecordSuffixes(store, toAddr, fromAddrs)
+	var rv []*quarantine.QuarantineRecord
+	for _, suffix := range allSuffixes {
+		key := quarantine.CreateRecordKey(toAddr, suffix)
+		if store.Has(key) {
+			bz := store.Get(key)
+			rv = append(rv, k.mustBzToQuarantineRecord(bz))
+		}
+	}
+	return rv
+}
+
+// AddQuarantinedCoins records that some new funds have been quarantined.
+func (k Keeper) AddQuarantinedCoins(ctx sdk.Context, coins sdk.Coins, toAddr sdk.AccAddress, fromAddrs ...sdk.AccAddress) error {
+	qr := k.GetQuarantineRecord(ctx, toAddr, fromAddrs...)
+	if qr != nil {
+		qr.AddCoins(coins...)
+	} else {
+		qr = &quarantine.QuarantineRecord{
+			Coins: coins,
+		}
+		for _, fromAddr := range fromAddrs {
+			if k.IsAutoAccept(ctx, toAddr, fromAddr) {
+				qr.AcceptedFromAddresses = append(qr.AcceptedFromAddresses, fromAddr)
+			} else {
+				qr.UnacceptedFromAddresses = append(qr.UnacceptedFromAddresses, fromAddr)
+			}
+		}
+	}
+	if qr.IsFullyAccepted() {
+		fromAddrStrs := make([]string, len(fromAddrs))
+		for i, addr := range fromAddrs {
+			fromAddrStrs[i] = addr.String()
+		}
+		return fmt.Errorf("cannot add quarantined funds %q to %s from %s: already fully accepted",
+			coins.String(), toAddr.String(), strings.Join(fromAddrStrs, ", "))
+	}
+	// Regardless of if its new or existing, set declined based on current auto-decline info.
+	qr.Declined = k.IsAutoDecline(ctx, toAddr, fromAddrs...)
+	k.SetQuarantineRecord(ctx, toAddr, qr)
+	return ctx.EventManager().EmitTypedEvent(&quarantine.EventFundsQuarantined{
+		ToAddress: toAddr.String(),
+		Coins:     coins,
+	})
+}
```

Comment:
Are you okay? That diff is unintelligible, and looks like it wouldn't even compile. Your diff also removes several key aspects of this function.

Reply:
Your suggestions don't set the `qr.AcceptedFromAddresses` or `qr.UnacceptedFromAddresses` field, which need to be set in here. Your suggestions also don't properly set `qr.Declined` (you just ignore the output of `IsAutoDecline`).

Reply:
I'm going to leave it as is.

* Your version also omits the addition of `coins` to an existing record.
* The function should not be returning anything except a possible error. Returning a `sdk.Coins` isn't helpful anywhere. And as you've got it, it's just returning an argument that was provided anyway.
* The loop cannot be removed from the `if qr. IsFullyAccepted()` block because the addresses need to be converted to strings for the error message.
* The `qr.IsFullyAccepted()` returns a bool already, so there's no need to assign it to a variable like that.
* The error in the `qr.IsFullyAccepted()` block is quite long, so it's nice to have it split on multiple lines.
* The `EmitTypedEvent` line is also long, and is better to split on multiple lines.
* The `SetQuarantineRecord` function does not return an error.

I think the only thing I like here is the combination of lines 253-255 into one line.

We generally try to keep lines under 120 characters. We often make exceptions for this, specially with function definitions. This isn't a hard-and-fast rule by any means, but in cases like the event and error lines, those can easily be split to multiple lines, improving readability.